File: Lua/interpreter.lua
Environment = { parent=nil, variables={}, functions={}, constants={} }
function Environment.new(parent)
    self.parent = parent
    self.functions = {
        ["helloCarlos"] = print
    }
    self.constants = {}
    self.variables = {}
    return self
end

function Environment.callvar(environment, name)
    local env = Environment.findhome(environment, name)
    if env ~= nil and env.variables ~= nil and env.variables[name] ~= nil then
        return env.variables[name]
    elseif env ~= nil and env.constants ~= nil and env.constants[name] ~= nil then
        return env.constants[name]
    else error("Variable "..name.." doesn't exist!") end
end

function Environment.declarevar(env, access, name, value)
    if access == "DeclareRock" then
        if env.variables ~= nil and env.variables[name] ~= nil then error("Variable "..name.." already declared!") end
        env.variables[name] = value
    elseif access == "ConstRock" then
        if env.constants ~= nil and env.constants[name] ~= nil then error("Variable "..name.." already declared!") end
        env.constants[name] = value
    else error("Can't declare variable with: "..access) end
    return value
end

function Environment.assignvar(environment, name, value)
    local env = environment.findhome(name)
    if env.constants ~= nil and env.constants[name] ~= nil then error("Can't assign constant variable "..name) end
    env.variables[name] = value
    return value
end

function Environment.callfunction(env, fname, fparams)
    if env.functions[fname] == nil then error("Can't find function "..fname) end
    if type(fparams) == "table" then
        for i = 1, #fparams do
            env.functions[fname](fparams[i])
        end
    else env.functions[fname](fparams)
    end
    return fname
end

function Environment.findhome(env, name)
    if env ~= nil and env.parent ~= nil and (env.parent.variables[name] ~= nil or env.parent.constants[name] ~= nil) then
        return env.parent
    elseif env ~= nil and (env.constants ~= nil and env.variables ~= nil) and env.variables[name] ~= nil or env.constants[name] ~= nil and env ~= nil then
        return env
    else error("Variable "..name.." doesn't exist!") end
end

function evaluate(exps)
    local env = Environment.new(nil)
    local result = 0
    for i=1, #exps do result = evaluateprimary(exps[i], env) end
    return result
end

function evaluateprimary(exp, env)
    if exp.node == "NumberLiteral" or exp.node == "StringLiteral" then
        return exp.left
    elseif exp.node == "Identifier" then
        return Environment.callvar(env, exp.left)
    elseif exp.node == "DeclareExpression" then
        return Environment.declarevar(env, exp.addon, exp.left.left, evaluateprimary(exp.right))
    elseif exp.node == "AssignExpression" then
        return Environment.assignvar(env, exp.left, evaluateprimary(exp.right))
    elseif exp.node == "FunctionCall" then
        return evaluatefunctioncall(exp, env)
    elseif exp.node == "BinaryExpression" then
        return evaluatebinary(exp, env)
    else error("Can't evaluate expression of node: "..exp.node) end
end

function evaluatebinary(exp, env)
    local left = evaluateprimary(exp.left, env)
    local right = evaluateprimary(exp.right, env)
    if type(left) == "number" and type(right) == "number" and exp.operator == "PlusRock" then
        return left + right
    elseif type(left) == "number" and type(right) == "number" and exp.operator == "MinusRock" then
        return left - right
    elseif type(left) == "number" and type(right) == "number" and exp.operator == "MultRock" then
        return left * right
    elseif type(left) == "number" and type(right) == "number" and exp.operator == "DivRock" then
        return left / right
    elseif type(left) == "number" and type(right) == "number" and exp.operator == "ModRock" then
        return left % right
    elseif type(left) == "string" and type(right) == "string" and exp.operator == "PlusRock" then
        return left .. right
    elseif type(left) == "string" and type(right) == "string" and exp.operator == "MinusRock" then
        return left:gsub(right, "")
    else error("Can't evaluate binary expression") end
end

function evaluatefunctioncall(exp, env)
    local fparams = {}
    if type(exp.right) == "table" then for i = 1, #exp.right do fparams[i] = evaluateprimary(exp.right[i], env) end
    else error("Can't parse function arguments") end
    return Environment.callfunction(env, exp.left.left, fparams)
end

File: TypeScript/interpreter.ts
import { NodeRole, Program, Statement, Expression, BinaryExpression, Identifier, NumericLiteral, NullVoid, VarDeclaration, VarAssignment, StringLiteral, FunctionCall } from "./ast.ts";
import { Role, KeyWords } from "./lexer.ts";
import Ecosystem from './environment.ts';

export type ValueRole = "null"
| "number"
| "bool"
| "identifier"
| "string";

export interface RuntimeResult {
    type: ValueRole,
}

export interface NullResult extends RuntimeResult {
    type: "null",
    value: "null"
}

export interface NumberResult extends RuntimeResult {
    type: "number",
    value: number
}

export interface StringResult extends RuntimeResult {
    type: "string",
    value: string
}

export interface BooleanResult extends RuntimeResult {
    type: "bool",
    value: boolean
}

export function evaluateProgram(program: Program, eco: Ecosystem): RuntimeResult {
    let currentresult: RuntimeResult = { type: "null", value: "null"} as NullResult;
    for(const statement of program.body) currentresult = evaluateStatement(statement, eco);
    return currentresult;
}

export function evaluateBinaryExp(statement: BinaryExpression, eco: Ecosystem): RuntimeResult {
    const left = evaluateStatement(statement.left, eco) as NumberResult;
    const right = evaluateStatement(statement.right, eco) as NumberResult;
    let result: number = 0;
    if(left.type == "number" && right.type == "number") {
        switch(KeyWords[statement.operator]) {
            case Role.PlusOperator: 
                result = left.value + right.value;
                break;
            case Role.MinusOperator: 
                result = left.value - right.value;
                break;
            case Role.MultOperator:
                result = left.value * right.value;
                break;
            case Role.DivOperator:
                result = left.value / right.value;
                break;
            case Role.ModOperator:
                result = left.value % right.value;
                break;
        }
        return { type: "number", value: result } as NumberResult;
    }
    return { type: "null", value: "null"} as NullResult;
}

export function evaluateVarDeclaration(statement: VarDeclaration, eco: Ecosystem): RuntimeResult {
    let value: RuntimeResult = evaluateStatement(statement.value, eco);
    switch(statement.constant) {
        case true:
            return eco.declareConst(statement.identifier, value);
        case false:
            return eco.declareVar(statement.identifier, value);
    }
}

export function evaluateVarAssignment(statement: VarAssignment, eco: Ecosystem): RuntimeResult {
    let value: RuntimeResult = evaluateStatement(statement.value, eco);
    return eco.assignVar(statement.identifier, value);
}

export function evaluateFunctionCall(statement: FunctionCall, eco: Ecosystem): RuntimeResult {
    let args: RuntimeResult[] = [];
    while(statement.fparams.length > 0) args.push(evaluateStatement(statement.fparams.shift() as Statement, eco));
    return eco.callFunc(statement.fname, args);
}

export function evaluateStatement(statement: Statement, eco: Ecosystem): RuntimeResult {
    switch(statement.kind) {
        case NodeRole.StringLiteral:
            return {
                type: "string",
                value: (statement as StringLiteral).string
            } as StringResult;
        case NodeRole.Identifier:
            const stmt: Identifier = statement as Identifier;
            return eco.callVar(stmt.string);
        case NodeRole.NumericLiteral:
            return {
                type: "number",
                value: (statement as NumericLiteral).number
            } as NumberResult;
        case NodeRole.FunctionCall:
            return evaluateFunctionCall(statement as FunctionCall, eco);
        case NodeRole.NullLiteral:
            return {
                type: "null",
                value: "null"
            } as NullResult;
        case NodeRole.BinaryExpression:
            return evaluateBinaryExp(statement as BinaryExpression, eco);
        case NodeRole.VarDeclaration:
            return evaluateVarDeclaration(statement as VarDeclaration, eco);
        case NodeRole.VarAssignment:
            return evaluateVarAssignment(statement as VarAssignment, eco);
        default:
            console.error(`Statement cannot be interpreted atm, Sorry!: ${statement.kind}`);
            process.exit();
    }
}